blocks: Port over some of the type information for @wordpress/blocks from DefinitelyTyped (#77393)
WordPress/gutenberg: packages/blocks/CHANGELOG.md

## Unreleased

> - `registerBlockType`: Introduced a slightly modified version of the generics from `@types/wordpress__blocks`
> - `registerBlockVariation`, `unregisterBlockVariation` updated the types to match the dispatch call

Reviewer: ```suggestion
- Update `registerBlockType` to accept the attributes type as a generic.
- Update `registerBlockVariation`, `unregisterBlockVariation` type signature to match the dispatch call.
```

Reviewer: Sorry, I missed this 🤦. Let us create a follow up to update the changelog entry.

Author: I made a PR to update the changelog: #77437
